docs/src/proposals/interest-bearing-tokens.md

```diff
@@ -0,0 +1,284 @@
+during the `GetHoldingBalance` instruction, the program needs to know the current
+time in order to properly convert from token amount to share amount.
+
+Without dynamic sysvars, we need to create a system to tell what sysvars are
+needed for different instructions on each program, which creates complications
+for all users.
+
+### RPC
+
+The validator's RPC server needs to properly handle new token program accounts
+for read-only instructions and pre / post balances.
+
+#### `getBalance` and `getSupply`
+
+Instead of deserializing SPL token accounts and returning the data, the
+RPC server runs read-only transactions on the i-token program.
+
+RPC simply uses the same flow as on-chain programs: create an ephemeral account
+then runs the read-only transaction onto the bank requested.
+
+TODO how do we set the cost of read-only transactions? Do we lower it specially
+for read-only RPC calls? The i-token program will probably be one of the more
+cost-intensive computations possible since it is performing present value
+discounting.
+
+#### `preTokenBalances` and `postTokenBalances`
+
+As mentioned earlier in the Token Program Conformance section, the i-token
+program caches balances in the `amount` field of the SPL token holding. This
+should only be used by RPC, and not by programs to get the balance of a non-SPL
+holding.
+
+### Associated Token Program
+
+The Associated Token Program needs to support all token programs seamlessly.
+Currently, it performs the following sequence:
+
+* `transfer` required lamports
+* `allocate` space
+* `assign` to the SPL token program
+* call `InitializeAccount` on SPL token
+
+This does not work for an opaque token program, because we do not know the size required
+from the outside. Conversely, if we allow a token program to take the lamports it wants,
+a malicious token program could take too much from users.
+
+To get around this, the Token Program interface has a new `CreateHolding`
+instruction, which only allocates space and assigns the account to the program.
+Once the space is allocated, the Associated Token Program can transfer the
+required lamports, and return an error if that number is too large. The process
+becomes:
+
+* call `CreateHolding` on the token program (allocate and assign)
+* calculate rent requirement based on the data size of holding account
+* `transfer` required lamports
+* call `InitializeAccount` on the token program
+
+### Web3 / Wallets
+
+To properly support i-tokens, wallets and applications will need to make the
+following changes:
+
+* get holdings: query for holdings owned by the public key, for all official
+token programs listed in the registry (not just `Tokenkeg...`)
+* get balances: avoid deserializing holding data, and instead use
+`getBalance` from RPC
+* transfer tokens: only use `ApproveChecked` and `TransferChecked` on the token
+program
+* create holding: avoid directly creating an account and initializing,
+and use the associated token account program instead
+
```

Comment:
our of curiosity, I wonder how ledger support aTokens? Maybe, we should follow the precedent?

Reply:
Hmm, shouldn't the Ledger display token amount, not share?
Based on the description in the Proposed Solution section, I understood that the program would take care of all conversions to shares, thus the instruction data passed in (and displayed on the Ledger) would be in tokens.

Reply:
> our of curiosity, I wonder how ledger support aTokens? Maybe, we should follow the precedent?

AFAIK, it doesn't. Just standard ERC20

Reply:
> Hmm, shouldn't the Ledger display token amount, not share?
> Based on the description in the Proposed Solution section, I understood that the program would take care of all conversions to shares, thus the instruction data passed in (and displayed on the Ledger) would be in tokens.

All instructions process `shares` -- token amount is the `uiAmount`, which may now be a more involved computation than multiplying in the number of decimals places.
The Ledger only gets `shares` since it can't execute the code provided by the program to compute a `uiAmount` based on the current start of the chain

Reply:
I'm very confused, then. I thought converting back from tokens to current shares -- with whatever involved logic -- is what the program does when processing the instruction. The Ledger only has the instruction data to work with.

> Yes, the idea is that all interactions from the outside are in terms of tokens, and not shares.

Ref Jon's comment: https://github.com/solana-labs/solana/pull/15927/files#r595974135